File: 2022/2/algorism/ZIMBABWE.py
import sys
sys.setrecursionlimit(200000000)

def countResult(e: str, m: str, c: list[int], b:str):
    result = 0
    for i, v in enumerate(c):
        if v != 0:
            if b == '' and v == 0: continue

            c[i] -= 1
            b += str(i)

            r = countResult(e, m, c, b)
            if r == -1:
                return result % 1000000007
            result += r

            b = b[:-1]
            c[i] += 1
        if i == 9 and len(e) == len(b):
            if int(e) > int(b):
                if (int(b) % int(m)) == 0:
                    return 1
            else:
                return -1
    return result % 1000000007

def countNumber(e):
    result = [0 for i in range(10)]
    for i in e:
        result[int(i)] += 1
    return result

for i in range(int(input())):
    e, m = input().split()
    c = countNumber(e)
    r = countResult(e, m, c, "")
    print(r)

File: 2022/2/algorism/ZIMBABWE.js
const readline = require('readline');

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

var inputs = [];
let count;

const init = (input) => {
    let e = input[0], m  = input[1]
    let c = countPerNumber(e);
    let r = countPerResult(e, m ,c, "");
}

const countPerNumber = (e) => {
    let counts = [];
    e.toString().split('').forEach(num => {
        if(counts[+num] === undefined) counts[+num] = 0;
        counts[+num] += 1;
    });
    return counts;
}

const countPerResult = (e, m, c, b) => {
    let result = 0;
    c.forEach((num, index) => {
        if (num !== 0) {
            c[index] -= 1;
            b += index;
            result += countPerResult(e, m, c, b);
            return result;
        }
        if (index = 10) {
            console.log(b);
            if ((+b < +e)||(+b / +c))
                return 1;
        }
    });
}

rl.on('line', (line) => {
    if(count === undefined) {
        count = +line;
    } else {
        inputs.push(line.trim().split(' '));
        count--;
    }
    if(count === 0) rl.close();
}).on('close', () => {
    inputs = inputs.reverse();
    let input;
    while(input = inputs.pop()) {
        init(input);
    }
});